text/0000-subslice-offset.md

```diff
@@ -0,0 +1,344 @@
+
+## range_of
+
+```rust
+impl<T> [T] {
+    pub fn range_of(&self, subslice: &[T]) -> Option<Range<usize>>
+}
+```
+
+Returns the range referred to by a subslice.
+
+If the given slice falls entirely within this slice, returns the range
+of indices the subslice refers to. If the given slice is not a subslice
+of this slice, `None` is returned.
+
+If a range is returned, both ends are less than or equal to `self.len()`.
+
+Note that this does not look at the contents of the slice, but only at
+the memory addresses.
+
+```rust
+let a = [0; 5];
+
+assert_eq!(a.range_of(&a[2..5]), Some(2..5));
+assert_eq!(a.range_of(&[7, 8, 9]), None);
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The functions below are added to `impl<T> [T] { .. }` in `src/libcore/slice/mod.rs`,
+with doc comments based on the [Guide-level explanation][guide-level-explanation] above.
+
+```rust
+    pub fn index_of(&self, element: &T) -> Option<usize> {
+        let element = element as *const _;
+        let range = self.as_ptr_range();
+        if range.contains(&element) {
+            unsafe { Some(element.offset_from(range.start) as usize) }
+        } else {
+            None
+        }
+    }
+
+    pub fn range_of(&self, subslice: &[T]) -> Option<Range<usize>> {
+        let range = self.as_ptr_range();
+        let subrange = subslice.as_ptr_range();
+        if subrange.start >= range.start && subrange.end <= range.end {
+            unsafe {
+                Some(Range {
+                    start: subrange.start.offset_from(range.start) as usize,
+                    end: subrange.end.offset_from(range.start) as usize,
+                })
+            }
+        } else {
+            None
+        }
+    }
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- It might encourage users to use this in places where better alternatives
+  are available. For example: `.iter().find()` in combination with `.index_of()`
+  instead of `.iter().position()`.
+
+- In most parts of the library, two references to different but equal objects
+  behave the same. The memory address itself being significant instead of the
+  value it refers to is somewhat uncommon.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This functionality is already available through calculations with raw pointers.
+Often, people prefer to avoid that, which in many cases leads to complicating
+the code in other ways.
+
+The main consideration is which things should have a safe and ergonomic option,
+and which are best left to raw pointer manipulation.
+
+The proposed functions put the line between references and pointers:
+the functions only work on references; as soon as
+you're using raw pointers, you'll have to do things manually instead.
+
+The alternative is to try to cover some cases for raw pointers as well:
+
+## Raw Pointers
+
+By requiring a `&T` or a `&[T]` in these functions, we needlessly restrict them
```

Comment:
This restriction to dereferenceable pointers side-steps potential problems with pointer comparisons: compiler optimizations can change the results of pointer comparisons to be different from the order of the runtime addresses when the pointers are dangling. See for example section 4.6 of [this paper](https://people.mpi-sws.org/~jung/twinsem/twinsem.pdf).

It is not immediately apparent to what degree LLVM performs such optimizations, but it seems risky to assume that it never will. LLVM docs say [`icmp` on pointers works "as if [the pointers] were integers"](https://llvm.org/docs/LangRef.html#icmp-instruction) but this is imprecise and apparently contradicted by the fact that [some comparisons](https://godbolt.org/z/NWGoGt) which can be true at runtime under -O0 are [folded to `false` at -O2](https://godbolt.org/z/x50w1j).

cc @rust-lang/wg-unsafe-code-guidelines

Reply:
Thanks for the feedback!

I mention somthing similar briefly, but it's a bit hidden at the end of 'option 4':

>  The downside is that we lose the guarantees of a &[T], and can no longer make assumptions such as start <= end or about the maximum size of a slice (which is needed to safely use pointer::offset_from).

It's part of the reason why I'm proposing the versions with references, and only mention the pointers as alternatives. I should probably mention this more clearly right at the start of the raw pointers section.

Reply:
I did miss that part, thanks for pointing it out.

Those problems you list seem of a different nature, though. Not knowing that `start <= end` and that the slice spans no more than `isize::MAX` bytes requires a different, slightly less efficient implementation (e.g., some more comparisons). In contrast, it is not clear to me whether the problem of non-deterministic pointer comparison can be circumvented at all, since e.g. ptr->int casts [do not currently stop the LLVM optimization I showcased earlier](https://godbolt.org/z/ur5Gjt). That may have to change anyway to make LLVM's memory model consistent, but at present I don't have confidence the raw pointer version can be implemented correctly, at least according to the naive specification of comparing the pointer addresses numerically.

Reply:
@rkruppe unfortunately I don't think restricting to dereferencable pointers entirely side-steps the problems here when ZST are involved. ZST pointers are nominally "dangling" as far as LLVM is concerned (or at least, they might be). But then, as @HeroicKatora points out, there are larger problems with ZST.

Reply:
> That may have to change anyway to make LLVM's memory model consistent, but at present I don't have confidence the raw pointer version can be implemented correctly, at least according to the naive specification of comparing the pointer addresses numerically.

Indeed LLVM is currently in a sad state where it is not possible to reliably compare the integer addresses of two pointers. Casting to integers first and then comparing *should* work but LLVM incorrectly "optimizes" that to comparing at pointer type instead. We could conceivably hack around that by "obfuscating" the comparison enough, which however would likely be catastrophic for performance. Also see these LLVM bugs: [[1]](https://bugs.llvm.org/show_bug.cgi?id=34548), [[2]](https://bugs.llvm.org/show_bug.cgi?id=35229).